### ISSUE TITLE
Missing argument is not reported if other argument contains a type error

### ISSUE THREAD

**Issue by JelleZijlstra** [microsoft/pyright#8026]
Consider this code:

```python
def f(a: int, b: str): pass

f(
    a=(
        int([])  # type: ignore
    )
)
```

https://pyright-play.net/?strict=true&code=CYUwZgBGAUCGBcECWA7ALgGggI0QZzQCcBKRAB1jzwChqZqJGJYBeaBpz1NaAbQF1ijAMQQ0ATzIhESAOYoA9oRAdGxauqA

Pyright reports no errors here, even though obviously the `b` argument is missing. If you remove the `# type: ignore`, it only reports an error about the `int([])` call, but not about the missing `b` argument.

It appears that if any other argument in a call contains a type error, then missing arguments in the rest of the call are not reported.

The use case where I ran into this was constructing a class with a number of complex arguments. There was an error deep in one argument that I wanted to ignore, but pyright failed to report that I was missing a different argument at the top level.

**Comment by erictraut** [comment]
Hmm, I'm not able to repro the problem. If I remove the `# type: ignore` in your example, pyright generates two errors. The first of these two is: `Argument missing for parameter "b"`. 

https://pyright-play.net/?strict=true&code=CYUwZgBGAUCGBcECWA7ALgGggI0QZzQCcBKRAB1jzwChqZqJGJYBeaBpz1NaAbQF1iHRkKFA

**Comment by erictraut** [comment]
The reason that `# type: ignore` suppresses both errors in this case is because the `Argument missing for parameter "b"` error is associated with the entire call expression, so it spans multiple lines. Its line range intersects with the line that contains the `# type: ignore` comment, so it is suppressed. I consider that behavior to be correct.

**Comment by JelleZijlstra** [comment]
Hm, you're right. Maybe I incorrectly minimized my original case; I'll see if I can reproduce this with a self-contained sample.

**Comment by JelleZijlstra** [comment]
Here's a sample that does reproduce my problem:

```python
from dataclasses import dataclass


@dataclass
class A:
    f1: int
    f2: list[int]

@dataclass
class B:
    f3: list[int]

@dataclass
class C:
    f4: B
    f5: str


def run():
    x = C(
        f4=B(
            [
                A(
                    f2=[],
                )
            ]
        ),
    )
    return x
```

https://pyright-play.net/?strict=true&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sAoOgAXiTUwxpaygEEAuGqAVGABGHgQB2CfoOAAmManwYEAbXySAugyZVW7angBCfQUIDMCpavUItNRpQ5snUAMImZAFjGHpA4ACsYsogdDQw2MBQIACu4gAUAJQeAgAeUAC8bvF%2BpkKeGYY5eSVQKrmleVzFlbX%2BshkqGgA0FXVQiW0lWpWJraadpiDYCDEg4lCpQA

Here pyright reports the inner type error (`A` is missing the `f1` argument), but not the outer one (`C` is missing `f5`).

**Comment by JelleZijlstra** [comment]
A simpler version:

```python
from dataclasses import dataclass


@dataclass
class B:
    f3: str

@dataclass
class C:
    f4: B


def run():
    x = C(
        f5=B(1),
    )
    return x
```

https://pyright-play.net/?strict=true&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sAoOgAXiTUwxpaygCEAuGqAVGABmHlAwIQDJlVbtqeAMJ9BQgCxiudGjGzAoIAK4A7ABQBKFYIAeUALxRFp-qsHAArHa6mAjOYA0LgLmQQbYCIYgxlDWQA

This reports that I passed the wrong argument type to `B`, but not that `C` doesn't take an `f5` argument.

**Comment by erictraut** [comment]
Here's what's going on in this case. To make this clearer, I'm going to remove the `@dataclass` decorator from the sample and provide the synthesized `__init__` and `__new__` methods in their explicit forms.

```python
from typing import Any, Self

class B:
    def __new__(cls, *args: Any, **kwargs: Any) -> Self: ...
    def __init__(self, f3: str) -> None: ...

class C:
    def __new__(cls, *args: Any, **kwargs: Any) -> Self: ...
    def __init__(self, f4: B) -> None: ...

def run():
    x = C(
        f5=B(1),
    )
    return x
```

When evaluating the constructor call for `C`, pyright first evaluates the `__new__` call (in conformance with the recently updated chapter on [constructors](https://typing.readthedocs.io/en/latest/spec/constructors.html) in the typing spec). The call to `__new__` must be evaluated first because it can potentially return a type other than `C`, and that can influence the evaluation of the `__init__` method.

In this case, the evaluated call to `C.__new__` results in a type error. That means its evaluated return type can't be trusted. Pyright effectively treats it as an `Unknown` at this point. Because the result of the `__new__` method call is unknown, evaluation of the constructor call is terminated at that point. There's no way to know what class's `__init__` method should be used.

So, I think pyright's behavior is justifiable in this case and in keeping with the general philosophy of trying to avoid cascades of errors caused by one type error. I can point to other situations where suppressing a type error with a `# type: ignore` comment (as opposed to fixing type violation in the code) has downstream consequences.

This is also a special case in that it involves constructor calls, which are complicated by the fact that they involve a chain of type evaluations. Mypy doesn't run into this edge case currently because it does not handle constructor calls in a way that conforms to the typing spec.

I'll think about this a bit more, but my inclination is to close this as "by design".

**Comment by JelleZijlstra** [comment]
I see. Intuitively, I feel a type checker should first evaluate all the arguments, and only then start looking at `__new__` and `__init__`. In this case, it would detect an error in the argument and therefore infer a type of `Any`, but it should still be able to proceed and report errors about missing arguments. I understand bidirectional inference may make the story more complicated, though.

**Comment by erictraut** [comment]
> should first evaluate all the arguments

Yeah, that's not really possible. Evaluating the arguments requires an inference context (bidirectional type inference), and the context comes from the `__new__` and `__init__` signatures, which can be different.

I've done an experiment where I changed pyright's behavior to ignore type errors when evaluating `__new__` and proceeding as though there was no `__new__` present. I was curious what I'd see from mypy_primer. Here are the [results](https://github.com/microsoft/pyright/pull/8028). It's not bad, so I think I may go ahead with this change.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.366.
